app/src/ui/stash-changes/stash-and-switch-branch-dialog.tsx

```diff
@@ -0,0 +1,171 @@
+import * as React from 'react'
+import { Dialog, DialogContent, DialogFooter } from '../dialog'
+import { Repository } from '../../models/repository'
+import { Dispatcher } from '../dispatcher'
+import { VerticalSegmentedControl } from '../lib/vertical-segmented-control'
+import { Row } from '../lib/row'
+import { Branch } from '../../models/branch'
+import { ButtonGroup } from '../lib/button-group'
+import { Button } from '../lib/button'
+import { UncommittedChangesStrategy } from '../../models/uncommitted-changes-strategy'
+import { Octicon, OcticonSymbol } from '../octicons'
+import { PopupType } from '../../models/popup'
+
+enum StashAction {
+  StashOnCurrentBranch,
+  MoveToNewBranch,
+}
+interface ISwitchBranchProps {
+  readonly repository: Repository
+  readonly dispatcher: Dispatcher
+  readonly currentBranch: Branch
+
+  /** The branch to checkout after the user selects a stash action */
+  readonly branchToCheckout: Branch
+
+  /** Whether `currentBranch` has an existing stash association */
+  readonly hasAssociatedStash: boolean
+  readonly onDismissed: () => void
+}
+
+interface ISwitchBranchState {
+  readonly isStashingChanges: boolean
+  readonly selectedStashAction: StashAction
+}
+
+export class StashAndSwitchBranch extends React.Component<
+  ISwitchBranchProps,
+  ISwitchBranchState
+> {
+  public constructor(props: ISwitchBranchProps) {
+    super(props)
+
+    this.state = {
+      isStashingChanges: false,
+      selectedStashAction: StashAction.StashOnCurrentBranch,
+    }
+  }
+
+  public render() {
+    const { isStashingChanges } = this.state
+    return (
+      <Dialog
+        id="stash-changes"
+        title={__DARWIN__ ? 'Switch Branch' : 'Switch branch'}
+        onSubmit={this.onSubmit}
+        onDismissed={this.props.onDismissed}
+        loading={isStashingChanges}
+        disabled={isStashingChanges}
+      >
+        <DialogContent className="dialog-content">
```

Comment:
See https://github.com/desktop/desktop/pull/7230#discussion_r274384004 and https://github.com/desktop/desktop/pull/7230#discussion_r274475468. 

```suggestion
        <DialogContent>
```